### ISSUE TITLE
Bubblewrap not working in strata other than init

### ISSUE THREAD

**Issue by ethan2-0** [bedrocklinux/bedrocklinux-userland#245]
Bubblewrap errors when run in a stratum other than the stratum that provides init. Error message is at the bottom of the steps to reproduce. It's not clear to me why this is happening, though looking at the output of `strace`, it seems I'm getting `EPERM` on a `clone` syscall with `flags=CLONE_NEWNS|CLONE_NEWUSER|SIGCHLD`, which makes sense given the error message.

To reproduce:

In my case, my init strat is named `debian`, using Debian stable, and I've also created `test-strat`, also Debian stable. Both have `bubblewrap` installed.

```sh
$ brl which
debian
$ brl deref init
debian
$ bwrap --version
bubblewrap 0.4.1
$ brl version
Bedrock Linux 0.7.24 Poki
$ bwrap --dev-bind / / echo hi
hi
$ strat test-strat
$ bwrap --dev-bind / / echo hi
bwrap: No permissions to create new namespace, likely because the kernel does not allow non-privileged user namespaces. See <https://deb.li/bubblewrap> or <file:///usr/share/doc/bubblewrap/README.Debian.gz>.
```

**Comment by cptpcrd** [comment]
Hmm, a quick look at the `clone(2)` man page shows this:

> EPERM (since Linux 3.9)
> CLONE_NEWUSER  was specified in the flags mask and the caller is in a chroot environment (i.e., the caller's root directory does not match the root directory of the mount namespace in which it resides).

This appears to be a security feature, likely to prevent unprivileged users from using user namespaces to escape chroots. Since Bedrock uses chroots to run non-init strata, this prevents creating user namespaces outside of the init strata.

This [check](https://github.com/torvalds/linux/blob/master/kernel/user_namespace.c#L92) also doesn't appear to have any exceptions. It *might* be possible to work around it by creating a new mount namespace when switching strata, but 1) I'm not 100% sure that would work and 2) if Bedrock did that by default, it would probably break things.

**Comment by paradigm** [comment]
I agree that `clone(2)` `EPERM` item is likely the culprit.  I also agree that per-stratum mount namespaces would likely fix this issue.

In the immediate future, work-arounds include:

1. Manually running `chmod u+s /path/to/bwrap` as root.  AFAIK `bwrap` is designed to be run as `setuid` in case the kernel has non-privileged user namespaces disabled.  However, I certainly understand unnecessary `setuid` being undesirable.
2. Pairing init with `bwrap`.  This constraint is undesirable as well.

Bedrock 0.7.x relies on the common mount namespace pervasively.  A ready example is `brl which` which compares PID 1's mount table against another PID's to determine which stratum provides the second PID.  I don't think a quick fix via a point update is viable.

The design of the upcoming 0.8.x is still somewhat fluid.  I can try to incorporate per-stratum mount namespaces into it, although I can't make any promises.  Off the top of my head it may introduce some design regressions:

- `strat` currently only requires `CAP_SYS_CHROOT`.  This will likely require `strat` be full blown `setuid`.
- This will effectively bump the Bedrock minimum kernel version.
- 0.7 requires a system-wide daemon ("crossfs") and a per-stratum daemon ("etcfs").  I was hoping 0.8 would drop that to just system-wide daemon ("brld") with at most one idle process/thread.  It's not clear to me if this could still be met; it may require, at a minimum, a per-stratum thread to hold open / control the namespace.
- Creating new global directories (equivalent of `bedrock.conf`'s `share = ` lines) will probably require a reboot.  AFAIK it's not possible to create new bind mounts across mount namespaces outside of a shared subtree mount.

While in principle having `bwrap` from any stratum just work is certainly desirable, it's not obvious to me if these trade-offs are worthwhile.  It's also not obvious to me that it's *not*.  I'll need to think about it.

**Comment by paradigm** [comment]
While not the main focus of this issue, I should point out that neither querying for the current shell then running a command like so:

> ```shell
> $ brl which
> debian
> [...]
> $ bwrap --dev-bind / / echo hi
> hi
> ```

nor specifying a shell then running a command like so:

> ```shell
> $ strat test-strat
> $ bwrap --dev-bind / / echo hi
> bwrap: No permissions to create new namespace, likely because the kernel does not allow non-privileged user namespaces. See <https://deb.li/bubblewrap> or <file:///usr/share/doc/bubblewrap/README.Debian.gz>.
> ```

are guaranteed to get the command from the shell's stratum.  Consider what happens in both those cases if `bwrap` is installed in a third stratum and not in either of those strata.  Another example which may be easier to think about is:

```shell
$ strat debian
$ brl which
debian
$ grep "^NAME" /etc/os-release
NAME="Debian GNU/Linux"
$ pacman --help | head -n1
usage:  pacman <operation> [...]
```

Keep in mind that, despite the discussion around namespaces and chroot, Bedrock is _not_ containers.

Rather, I recommend either querying specifically about the command in question (rather than the shell):

```shell
$ brl which bwrap # if I run `bwrap` in this context, which stratum provides it?
debian
$ bwrap | head -n1
usage: bwrap [OPTIONS...] [--] COMMAND [ARGS...]
$ strat test-strat
$ brl which bwrap # if I run `bwrap` in this context, which stratum provides it?
test-strat
$ bwrap | head -n1
usage: bwrap [OPTIONS...] [--] COMMAND [ARGS...]
```

or just explicitly specifying which stratum's instance is desired:

```shell
$ strat debian bwrap | head -n1
usage: bwrap [OPTIONS...] [--] COMMAND [ARGS...]
$ strat test-strat bwrap | head -n1
usage: bwrap [OPTIONS...] [--] COMMAND [ARGS...]
```

**Comment by paradigm** [comment]
After thinking about this even more, I think per-stratum namespaces would also help with:

- [The GRUB+BTRFS/ZFS issue](https://github.com/bedrocklinux/bedrocklinux-userland/issues/200).  AFAIK that's because GRUB does not handle multiple instances of a `/boot` mount well.  If we namespace it, it'll only see one.
- `java` prints a warning about `cpuset` mounts, which have surprised some users.

The trade-off seems more and more in favor of per-stratum namespaces.  I'm going to start planning a big refactor of the 0.8 efforts in this direction.
